cmd/keeperd/main.go

```diff
@@ -0,0 +1,203 @@
+//
+// Copyright 2019 Insolar Technologies GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/cobra"
+	jww "github.com/spf13/jwalterweatherman"
+	"github.com/spf13/viper"
+
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/log"
+)
+
+var configPath string
+
+func main() {
+	var rootCmd = &cobra.Command{
+		Use: "keeperd --config=<path to config>",
+		Run: rootCommand,
+	}
+	rootCmd.Flags().StringVarP(&configPath, "config", "c", "", "path to config file")
+	err := rootCmd.Execute()
+	if err != nil {
+		fmt.Println("Wrong input params:", err.Error())
+	}
+}
+
+func rootCommand(cmd *cobra.Command, args []string) {
+	jww.SetStdoutThreshold(jww.LevelInfo)
+	var err error
+
+	vp := viper.New()
+	cfg := NewConfig()
+	if len(configPath) != 0 {
+		vp.SetConfigFile(configPath)
+	}
+	err = vp.ReadInConfig()
+	if err != nil {
+		log.Warn("failed to load configuration from file: ", err.Error())
+	}
+	err = vp.Unmarshal(&cfg)
+	if err != nil {
+		log.Warn("failed to load configuration from file: ", err.Error())
+	}
+
+	ctx := context.Background()
+	ctx, logger := inslogger.InitNodeLogger(ctx, cfg.Log, "main", "", "keeperd")
+
+	keeper := NewKeeper(cfg.Keeper)
+	keeper.Run(ctx)
+
+	vp.WatchConfig()
+	vp.OnConfigChange(func(e fsnotify.Event) {
+		logger.Info("Reloading config file")
+		cfg := NewConfig()
+		err := vp.Unmarshal(&cfg)
+		if err != nil {
+			logger.Errorf("Failed to reload config: %s", err.Error())
+			return
+		}
+		keeper.SetConfig(cfg.Keeper)
+	})
+
+	var gracefulStop = make(chan os.Signal, 1)
+	signal.Notify(gracefulStop, syscall.SIGTERM)
+	signal.Notify(gracefulStop, syscall.SIGINT)
+
+	<-gracefulStop
+}
+
+type Keeper struct {
+	config      KeeperConfig
+	isAvailable uint32
+}
+
+func NewKeeper(cfg KeeperConfig) *Keeper {
+	return &Keeper{
+		config:      cfg,
+		isAvailable: 0,
+	}
+}
+
+func (k *Keeper) SetConfig(cfg KeeperConfig) {
+	k.config = cfg
```

Comment:
There is a race condition here. SetConfig is not thread safe.

Reply:
Fixed